### ISSUE TITLE
Printer emulation cuts rightmost character

### ISSUE THREAD

**Issue by Schwefelholz** [GideonZ/1541ultimate#306]
Tried this in release 3.7 on Ultimate II, 3.10a on Ultimate 64 Elite and 3.10e on Ultimate II+L:

Using Commodore MPS, Epson FX-80/JX-80, IBM Graphics or IBM Proprinter emulations, once a printed line exceeds 80 characters of length, a character is cut from the right hand side. See attached printout: 1st line misses a colon, 2nd line missed an "I" on the right hand side.

![Page-001](https://user-images.githubusercontent.com/6893444/226280377-48d13788-21cf-4bd6-a2c7-38af1c1b1eb3.png)

**Comment by Schwefelholz** [comment]
Now that your write this: I've seen those one-pixel column characters as well in another printout.

Is this something that can make it into the next firmware? Any chance it can get fixed in the Ultimate II?

**Comment by rgc2000** [comment]
Yes, last official firmware for U2 is 3.7 and I can see that I am using a custom firmware 3.8 / FPGA 117 on my U2, not very official certainly unfinished work in progress inside. Version 3.9 does not load on U2, not enough room for secondary boot loader. Anyway, back porting the fix to 3.7 is easy.
Gideon, you are right the three targets build for U2, my local git was not up to date.

**Comment by Schwefelholz** [comment]
That would be very cool!
--
Sichere Kommunikation per Kurznachricht:
Threema: https://threema.ch/download
oder
Signal: http://sgnl.link/1CYCQQN

per E-Mail:
GnuPG-Key: gpg --recv-keys 0x29AA1668
GnuPG-Fingerprint: BDB7 63C2 8A48 A6A7 25FF  6D80 C444 F8C8 29AA 1668

--
"Those who would give up essential Liberty to purchase a little temporary Safety, deserve neither Liberty nor Safety." (Benjamin Franklin)

Am 21. März 2023 00:03:16 MEZ schrieb "René" ***@***.***>:

**Comment by Grrrolf** [comment]
> As far as I'm aware, the latest official firmware for U2 is 3.7.

The latest firmware can be found here: https://ultimate64.com/Firmware and it is version 3.10a. There are also U2 updaters in the zip file. Networking may be broken for the U2 in this firmware, according to reports in the FB group.

**Comment by rgc2000** [comment]
@Grrrolf , you are right. I confirm that 3.10a works on both U2 and U2+ but it looks like I have a memory stick issue on U2+ when printing several pages. Sometimes the USB device goes "FAILED" and then printer is stuck and PNG files may be lost. No problem at all on firmware 3.9 with the same memory stick. But that's not what this discussion is about.
And again : printer is so much faster on U2 than U2+, most of time is spent on PNG generation, in bitmap compression routine.

**Comment by daglem** [comment]
> And again : printer is so much faster on U2 than U2+, most of time is spent on PNG generation, in bitmap compression routine.

@GideonZ Perhaps it would be beneficial to switch to using Nios V or another RISC-V implementation for the U2+?
Ideally the U2 would use a RISC-V CPU as well, to streamline everything (IUC the U2+L will be using NEORV32).

**Comment by daglem** [comment]
@GideonZ Sounds good! Regarding capable open source RISC-V implementations, here is a (slightly dated) report on that, in case you want to look into it later: https://www.esa.informatik.tu-darmstadt.de/assets/publications/materials/2019/heinz_reconfig19.pdf

Given that you want a VHDL implementation of RISC-V, Orca is probably quite a bit faster than NEORV32. Size may be an issue though, and it doesn't seem to be actively developed:
https://github.com/vbx-aseverance/orca (and possibly other copies on GitHub).

Otherwise I know that several people are pleased with VexRiscv, which has a good performance/size ratio. This is a good bet to fit inside all the U2s, as it's very configurable. The downside (IMHO) is the dependency on Scala: https://github.com/SpinalHDL/VexRiscv

If you really want to fly, then you could possibly have a look at cva5 (aka Taiga in the report above). This is the largest of the three, and the tools you use may not be too happy about SystemVerilog: https://github.com/openhwgroup/cva5

Speaking of vendor lock-in, have you considered using GHDL + yosys and nextpnr for the U2+L, which I presume is Lattice ECP5?
Perhaps GHDL + yosys and nexpnr could be used for the U2 and/or the U2+ as well, I don't know the status for the FPGAs you use there (which are they?)

**Comment by rgc2000** [comment]
Considering the issue regarding the extra character cut in the right, this is solved in my branch. I am doing a merge request.

**Comment by GideonZ** [comment]
> @GideonZ I would be interested in trying out GHDL + yosys and nextpnr once you have something ready for the U2+L. 

@daglem : Not sure what you meant by "if you have something ready". The units are now shipping, so everything is in this repo, both FPGA target (u2plus_ecp5) as well as the software targets (target/software/riscv32*)

**Comment by Schwefelholz** [comment]
Great. Thank you!

**Comment by daglem** [comment]
> > @GideonZ I would be interested in trying out GHDL + yosys and nextpnr once you have something ready for the U2+L.
> 
> @daglem : Not sure what you meant by "if you have something ready". The units are now shipping, so everything is in this repo, both FPGA target (u2plus_ecp5) as well as the software targets (target/software/riscv32*)

I don't see any new firmware release, nor any recent tag, so I assumed it wasn't ready. Should I just use master, or did you use any particular branch / commit for the shipped U2+L firmware?

**Comment by GideonZ** [comment]
Master should  work. There is no tag yet, because the changes that I needed to make for U2+L broke the U2. As soon as I get U2 up and running again, I will release and add a tag.

**Comment by rgc2000** [comment]
> Platform	CPU	Frequency	Instruction Cache	Data Cache
> U2 >3.0	MBLite	50 MHz	Yes	Yes
> U2+	Nios-IIe	62.5 MHz	No	No
> U64	Nios-IIf	66.6 MHz	Yes	No
> U2+L	NeoRV32	50 MHz	Yes	No

Interesting, I should try the U64. I don't have one yet, mostly because I am too lazy to find a case, keyboard, brackets and so on to mount a new computer. I will probably order you one when available, but I am not in a hurry.

**Comment by markusC64** [comment]
> Master should work. There is no tag yet, because the changes that I needed to make for U2+L broke the U2. As soon as I get U2 up and running again, I will release and add a tag.

Well, I hope that includes fixing it on the U2+. 3.10e does not run nice on my C128 with the Ultimate II+ (not U2+L). While Firmware 3.10c runs perfectly nice on that system, with firmware 3.10e the C128 crashes while freezing even in 1 MHz mode.

**Comment by rgc2000** [comment]
New merge request #309 . Previous request contained unwanted VHDL code not related to the issue.

**Comment by rgc2000** [comment]
Forget merge request #309. Fix is also included in merge request #345 with much more other issues and enhancements.

**Comment by rgc2000** [comment]
Fixed in firmware 3.11
